### PR TITLE
LfMerge build produces container instead of .deb files

### LINKED ISSUE #185: Build should produce a container image instead of .deb packages
Basic outline, more of a rough draft, of how this could work:

- Take the debian/rules.in file and convert the "install" section to a shell script.
- Then in the Docker container process, run that to install things into a directory.
- Then copy the directory out to a tarball.
- Then there's another Dockerfile that builds the lfmerge image.
- It takes the tarball(s) from the build container(s) and puts them directly into the image with `ADD lfmerge-fw8.tar /`

### PR DESCRIPTION
**Pull request by rmunn**
Closes #185.

The basic way this works is that the `debian/rules` file (a Makefile that has to have specifically-named targets for the Debian build process to process it) is converted to a shell script. It puts the compilation results into place in a `tarball` directory, which is then copied into a new Docker image.

Roadmap for this PR before I mark it ready for review:

* [x] Test the built lfmerge container locally
* [x] Edit GHA workflow to push the built container to Docker Hub

---
This change is [Reviewable](https://reviewable.io/reviews/sillsdev/lfmerge/186)

### CODE REVIEW

**Review by longrunningprocess — CHANGES_REQUESTED** [review]
I'm really excited about this effort, it will be a really valuable refactor!

Most of my thoughts were relatively stylistic and some were questioning a possible oversight on your part but the workflow is really complex and I'd like to see some of those changes unless you want to discuss them further.

File: .github/workflows/docker-build.yml [hunk 1]
```diff
@@ -10,8 +10,35 @@ on:
   workflow_dispatch:
 
 jobs:
+  calculate_branches:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out current branch
+        # actions/checkout@v2.0.10 is commit ec3a7ce113134d7a93b817d10a8272cb61118579
+        uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579
```
[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:18]: With [many](https://github.com/faisalman/ua-parser-js/issues/536) different [cases](https://blog.sonatype.com/newly-found-npm-malware-mines-cryptocurrency-on-windows-linux-macos-devices) of [hijacked](https://heimdalsecurity.com/blog/coa-and-rc-npm-packages-hijacked/) NPM packages, we **absolutely should not** be trusting that Github Actions packages won't ever be hijacked either. In fact, I plan to submit a PR replacing all the version numbers in the web-languageforge workflows with commit hashes, too.

File: .github/workflows/docker-build.yml [hunk 2]
```diff
@@ -20,33 +47,33 @@ jobs:
...
           fetch-depth: 0  # All history for all tags and branches, since GitVersion needs that
 
       - name: Verify current branch
-        run: git branch --contains HEAD --format '%(refname)'
-        if: steps.should_run.outcome == 'success'
+        run: |
+          git branch --contains HEAD --format '%(refname)'
+          echo FW8 was "${FW8Branch}"
+          echo FW9 was "${FW9Branch}"
+        env:
```
[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:71]: Why pass them through env? Because if the branch name happens to contain a `$` or `|` or `>` character, or even a `"` or `'` character, direct text replacement like `echo FW8 was ${{ needs.calculate_branches.outputs.FW8Branch }}` would end up having unintended effects and likely causing the build to fail. Passing the data in through environment variables ensures that special characters will get quoted correctly. There's a mention of this somewhere in the Github Actions docs, but I couldn't find it off-hand. But that's the reason: so that if someone, whether naïve or malicious, creates a branch named `John's testing branch` or `| tee /dev/null` or something, it won't cause a syntax error in the build.

File: .github/workflows/docker-build.yml [hunk 3]
```diff
@@ -131,25 +158,23 @@ jobs:
...
       - name: Build DBVersion-specific Docker image
-        if: steps.should_run.outcome == 'success'
         # docker/build-push-action@v2.7.0 is commit a66e35b9cbcf4ad0ea91ffcaf7bbad63ad9e0229
         uses: docker/build-push-action@a66e35b9cbcf4ad0ea91ffcaf7bbad63ad9e0229
         with:
           push: false
           load: true
           tags: lfmerge-build-${{matrix.dbversion}}
           context: .
+          file: Dockerfile
```
[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:168]: The action was failing, which is why I added this, but it turned out to be failing for a different reason. I can remove this one.

File: .github/workflows/docker-build.yml [hunk 4]
```diff
@@ -160,60 +185,62 @@ jobs:
...
+          path: tarball
     outputs:
       MajorMinorPatch: ${{ steps.version.outputs.MajorMinorPatch }}
+      MsBuildVersion: ${{ steps.version.outputs.MsBuildVersion }}
 
   release:
     runs-on: ubuntu-latest
     needs: build
-    if: github.event_name == 'push' && (github.ref == 'refs/heads/live' || github.ref == 'refs/heads/fieldworks8-live')
+    # if: github.event_name == 'push' && (github.ref == 'refs/heads/live' || github.ref == 'refs/heads/fieldworks8-live')
```
[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:204]: Yes. Once I change the `push: false` input on the docker-build-push action, as described below, I will remove this line.

File: .github/workflows/docker-build.yml [hunk 5]
```diff
@@ -160,60 +185,62 @@ jobs:
...
-        generate_release_notes: true
+        push: false
+        load: true
+        tags: sillsdev/lfmerge:${{ needs.build.outputs.MsBuildVersion }}
+        context: .
+        file: Dockerfile.finalresult
+      # TODO: Set push to equal the value of (github.event_name == 'push' && github.ref == 'refs/heads/live')
+
+    - name: Show metadata from LfMerge image build step
+      run: echo "$METADATA"
```
[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:241]: With passing the values through env, this allows them to be quoted correctly. Direct text substitution into the `echo` command would be unpredictable since I don't know what special characters will end up being in the metadata.

File: Dockerfile.finalresult [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    adduser --quiet www-data fieldworks ; \
+	install -d -o www-data -g www-data -m 02775 /var/www/.local
+
+FROM lfmerge-base AS lfmerge
+
+# install LFMerge prerequisites
+# python - required by Mercurial, which is bundled in the LFMerge installation
+# rsyslog - lfmerge logs to rsyslog and expects this to exist
+# rsyslog customizations (imklog reads kernel messages, which isn't allowed or desired in Docker containers)
+# logrotate - TODO: is this required?
```
[the remark this answers is not in the document]
Comment (rmunn) [reply Dockerfile.finalresult:32]: Not yet; I plan to make that determination later, after I've merged the container build PR.

File: Dockerfile.finalresult [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ADD tarball/lfmerge-7000069/ /
+ADD tarball/lfmerge-7000070/ /
+ADD tarball/lfmerge-7000072/ /
+# Add more lines here if we add new DbVersions
+
+ADD tarball/lfmerge/ /
+
+RUN mkdir -m 02775 -p /var/lib/languageforge/lexicon/sendreceive/syncqueue && chown -R www-data:www-data /var/lib/languageforge
+
+# TODO: Turn this into environment variables, because we want to be able to just set env vars in k8s config and restart the container
```
[the remark this answers is not in the document]
Comment (rmunn) [reply Dockerfile.finalresult:53]: "this" is the sendreceive.conf file in the next line. Later on, you comment on wanting to replace it with env vars; I agree, as you can see from this comment.

File: docker/scripts/build-and-test.sh [hunk 8]
```diff
@@ -29,4 +32,6 @@ fi
 
 rm -rf "$SCRIPT_DIR"/data/php   # So it doesn't get into the .deb source package
 
-"$SCRIPT_DIR"/build-debpackages-combined.sh ${DbVersion}
+# "$SCRIPT_DIR"/build-debpackages-combined.sh ${DbVersion}
```
[the remark this answers is not in the document]
Comment (rmunn) [reply docker/scripts/build-and-test.sh:35]: Nope. Now that I know the container build is working, I can remove it.

File: docker/scripts/create-installation-tarball.sh [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+# Install wrapper scripts
+install -d ${COMMONDESTDIR}/usr/bin
+install -m 755 lfmerge ${COMMONDESTDIR}/usr/bin
+install -m 755 lfmergeqm ${COMMONDESTDIR}/usr/bin
+install -m 755 startlfmerge ${DBDESTDIR}/${LIB}
+# Install conf file
+install -d ${COMMONDESTDIR}/etc/languageforge/conf
+install -m 644 debian/sendreceive.conf ${COMMONDESTDIR}/etc/languageforge/conf
+# Create working directories
+mkdir -p ${COMMONDESTDIR}/var/lib/languageforge/lexicon/sendreceive/state
```
[the remark this answers is not in the document]
Comment (rmunn) [reply docker/scripts/create-installation-tarball.sh:74]: Yes, that would work better. The original lines came from a Makefile, where perhaps the `{ }` shell syntax wasn't available, hence the way it was written. I'll replace this with a single line (and also add `syncqueue`, which was missing from the original).

File: lfmerge.conf [hunk 10]
```diff
@@ -0,0 +1,10 @@
+# Configuration file for LfMerge
```
[the remark this answers is not in the document]
Comment (rmunn) [reply lfmerge.conf:1]: Agreed, which is why I wrote that comment in the Dockerfile about "replac[ing] this with environment variables". (The "this" being this file.)

File: lfmergeqm-background.sh [hunk 11]
```diff
@@ -0,0 +1,10 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (rmunn) [reply lfmergeqm-background.sh:1]: Good idea. After I merge this PR, I'll figure out how to do that.

File: lfmergeqm-looping.sh [hunk 12]
```diff
@@ -0,0 +1,9 @@
+#!/bin/sh
```
[the remark this answers is not in the document]
Comment (rmunn) [reply lfmergeqm-looping.sh:1]: inotifywait is using the exact same system as https://github.com/longrunningprocess/docker-sandbox/tree/main/fs-watcher (which uses `inotifyd`). And I'm having a bit of trouble picturing how the inotify process would kick off lfmerge if it was in a separate container from the lfmerge container. And anyway, the whole point of this PR is to make the lfmerge container be a stand-alone container with volume mapping. I'm not sure that moving one tiny piece of it into a different container is a win.

File: mercurial-cacerts.rc [hunk 13]
```diff
@@ -0,0 +1,4 @@
+[web]
```
[the remark this answers is not in the document]
Comment (rmunn) [reply mercurial-cacerts.rc:1]: This config file is normally created by the `mercurial` package, which we don't install in the stripped-down `lfmerge` container. But without this config file, Send/Receive fails because Mercurial says "I can't verify the HTTPS fingerprint coming from the server". This config file points to the Debian-based list of certs that includes, among others, the LetsEncrypt root certs, so that our hg-public server can be verified.

In other words, this is required for Send/Receive to work, and it's better to just add this one file than the whole `mercurial` package.

File: .github/workflows/docker-build.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:71]: Found it: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable (also read the whole article)

File: .github/workflows/docker-build.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:18]: Also, https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsuses says that "Using the commit SHA of a released action version is the safest for stability and security."

File: .github/workflows/docker-build.yml [hunk 14]
```diff
@@ -10,8 +10,35 @@ on:
...
+      - name: Check out current branch
+        # actions/checkout@v2.0.10 is commit ec3a7ce113134d7a93b817d10a8272cb61118579
+        uses: actions/checkout@ec3a7ce113134d7a93b817d10a8272cb61118579
+        with:
+          fetch-depth: 0  # All history for all tags and branches, since branch calculation step needs that
+
+      - name: Calculate branches to build
+        run: |
+          echo "${GITHUB_REF}"
+          ./calculate-branches.sh | grep FW.Branch >> "${GITHUB_ENV}"
```
[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:25]: When FW 10 comes out, I'll need to change these to FW08 and FW09 and make the regex `FW..Branch`, with two dots instead of one. But that won't happen for a few more years, and it's more likely that I'll change the build process again before that happens.

File: .github/workflows/docker-build.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (megahirt) [reply .github/workflows/docker-build.yml:18]: Robin and I discussed this in the office and we will compromise with pinning it to the specific version number v2.0.10

[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:18]: After discussing this with Chris, I'll switch the actions that are official GitHub actions (`actions/checkout` and `actions/(up|down)load-artifacts`) to use version numbers, because it's so extremely unlikely that the official GitHub `actions` repositories will ever be hijacked. I'll keep using commit SHAs for the `docker/foo` actions, because there, I don't know how many people have access to the `docker` org account and how quickly it would be noticed if someone hijacked it. But with the official Microsoft-owned `actions` org, I'll acknowledge that the commit SHAs are probably more paranoid than is strictly necessary.

File: .github/workflows/docker-build.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply .github/workflows/docker-build.yml:241]: https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-an-intermediate-environment-variable explains why I did this. I would keep this one as-is as far as the env vars go, except that I'll end up removing it because it's debugging output that I don't need anymore.

File: Dockerfile.finalresult (hunk 6 above)
[the remark this answers is not in the document]
Comment (rmunn) [reply Dockerfile.finalresult:32]: Talked to Tim Eves, who said "Keep it". Because while at the moment the logs never get that big because the container is frequently deleted and re-created (every time we push a new LF deployment), there will come a day when the site goes into maintenance mode for a while. And then the logs will just accumulate, and might end up eating up all the space allocation for the pod. Using `logrotate` (which might require adding a config file or two) will allow us to delete old logs automatically, to ensure that we don't fill up the container. So I'll remove this TODO comment and keep the `logrotate` package.

File: .github/workflows/docker-build.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply .github/workflows/docker-build.yml:68]: lots of good advice in there, thanks

File: .github/workflows/docker-build.yml [hunk 15]
```diff
@@ -10,8 +10,34 @@ on:
   workflow_dispatch:
```
Comment (longrunningprocess) [inline .github/workflows/docker-build.yml:10]: I'm not sure you'll need this capability, might be better to simply remove it.

File: .github/workflows/docker-build.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply .github/workflows/docker-build.yml:68]: based on the triggers, maybe there's a disconnect here between what can trigger this workflow and what might appear in the output of your script?  Is it because of something that happens in your script that you now don't trust it?

File: .github/workflows/docker-build.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply .github/workflows/docker-build.yml]: great, thanks

File: .github/workflows/docker-build.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply .github/workflows/docker-build.yml]: sounds good

File: .github/workflows/docker-build.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply .github/workflows/docker-build.yml]: ok

File: Dockerfile.finalresult (hunk 6 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply Dockerfile.finalresult]: ok

File: docker/scripts/build-and-test.sh (hunk 8 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply docker/scripts/build-and-test.sh:35]: great

File: lfmerge.conf (hunk 10 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply lfmerge.conf:1]: ok

File: lfmergeqm-background.sh (hunk 11 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply lfmergeqm-background.sh:1]: sounds like a good approach, maybe you could just capture it i another issue or however you're tracking your work.

File: lfmergeqm-looping.sh (hunk 12 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply lfmergeqm-looping.sh:1]: that reference wasn't as much about inotify but the shared volume approach.  It would kick off via the dropzone (a shared volume)...we can talk about it if that will help clarify the idea.  Either way, it's another one in the same category as the background housekeeping conversation above, another issue for another time.

File: mercurial-cacerts.rc (hunk 13 above)
[the remark this answers is not in the document]
Comment (longrunningprocess) [reply mercurial-cacerts.rc:1]: ok

File: Dockerfile.finalresult [hunk 16]
```diff
@@ -29,7 +29,7 @@ FROM lfmerge-base AS lfmerge
 # python - required by Mercurial, which is bundled in the LFMerge installation
 # rsyslog - lfmerge logs to rsyslog and expects this to exist
 # rsyslog customizations (imklog reads kernel messages, which isn't allowed or desired in Docker containers)
-# logrotate - TODO: is this required?
+# logrotate - ensure logs won't grow without bound, and that old logs get cleaned up automatically
```
Comment (longrunningprocess) [inline Dockerfile.finalresult:32]: thank you!

**Comment by longrunningprocess** [comment]
> @longrunningprocess I had a conversation in the office with @rmunn about his PR and your review. Thank you for your feedback and giving this review a go. In the interest of Robin moving forward on this project (there are a number of PRs waiting on this one) I agreed he should go ahead and merge without waiting for you to wake up and approve the PR.

understood
